apps/ltn/src/edit/one_ways.rs

```diff
@@ -61,8 +62,70 @@ pub fn handle_world_outcome(
                 // We always create it from-scratch when needed.
             });
 
+            app.session.edits.before_edit();
+
+            let r_edit = app.map.get_r_edit(r);
+            // Was the road originally like this? Use the original OSM tags to decide.
+            // TODO This'll break in the face of newer osm2streets transformations. But it's the
+            // same problem as EditRoad::get_orig_from_osm -- figure out a bigger solution later.
+            if r_edit == EditRoad::get_orig_from_osm(app.map.get_r(r), app.map.get_config()) {
+                app.session.edits.one_ways.remove(&r);
+            } else {
+                app.session.edits.one_ways.insert(r, r_edit);
+            }
+
+            // We don't need to call after_edit; no filter icons have changed
+
             EditOutcome::Transition(Transition::Recreate)
         }
         _ => EditOutcome::Nothing,
     }
 }
+
+// This is defined here because some of the heavy lifting deals with one-ways, but it might not
+// even undo that kind of change
+pub fn undo_proposal(ctx: &mut EventCtx, app: &mut App) {
+    let prev = app.session.edits.previous_version.take().unwrap();
+
+    // Generate edits to undo the one possible change to a one-way
+    if prev.one_ways != app.session.edits.one_ways {
+        let mut edits = app.map.get_edits().clone();
+
+        // Often the very last command in the edit stack is the one we need to undo. But when
+        // switching between different proposals, we accumulate more commands to sync map state, so
+        // this isn't true in general.
+
+        // So instead, look for the entry that's different
```

Comment:
This is a mess to reason about, but it holds up to my testing so far. Ideally we could just use `map_model::MapEdits`, which has a framework in place written by a much more level-headed Dustin for properly undoing/redoing commands, and even for transforming IDs to try to survive OSM changes. But that'll require a larger reckoning of modal filters into that layer, which I'm not prepared to tackle yet.